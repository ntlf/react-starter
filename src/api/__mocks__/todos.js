export const getTodos = async () => [
  {
    id: '1',
    text: 'One'
  },
  {
    id: '2',
    text: 'Two'
  }
];
