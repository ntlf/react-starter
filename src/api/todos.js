import axios from 'axios';

export const getTodos = async () => {
  const response = await axios.get(
    'https://wt-a339e6290d2b75021e179abd37d83555-0.sandbox.auth0-extend.com/react-ws/'
  );

  return response.data;
};

export const postTodo = async todo => {
  await axios.post(
    'https://wt-a339e6290d2b75021e179abd37d83555-0.sandbox.auth0-extend.com/react-ws/',
    todo
  );
};
