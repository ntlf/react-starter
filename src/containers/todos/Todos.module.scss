.Home {
  padding: 1em;
}
