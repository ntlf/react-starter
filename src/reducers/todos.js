import {
  FETCH_TODOS_REQUEST,
  FETCH_TODOS_SUCCESS,
  FETCH_TODOS_FAILURE
} from '../actions/types';

const initialState = {
  data: [],
  isFetching: true,
  error: ''
};

export default (state = initialState, action) => {
  switch (action.type) {
    case FETCH_TODOS_REQUEST:
      return {
        ...initialState,
        isFetching: true
      };
    case FETCH_TODOS_SUCCESS:
      return {
        ...state,
        isFetching: false,
        data: [...action.payload.data]
      };
    case FETCH_TODOS_FAILURE:
      return {
        ...state,
        isFetching: false,
        error: action.payload.error
      };
    default:
      return state;
  }
};
