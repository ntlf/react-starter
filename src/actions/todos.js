import {
  FETCH_TODOS_FAILURE,
  FETCH_TODOS_REQUEST,
  FETCH_TODOS_SUCCESS
} from './types';
import { getTodos, postTodo } from '../api/todos';

export const loadTodos = () => async dispatch => {
  dispatch({
    type: FETCH_TODOS_REQUEST
  });

  try {
    const todos = await getTodos();

    dispatch({
      type: FETCH_TODOS_SUCCESS,
      payload: { data: todos }
    });
  } catch ({ response: { data: error } }) {
    dispatch({
      type: FETCH_TODOS_FAILURE,
      payload: { error }
    });
  }
};

export const addTodo = todo => async dispatch => {
  await postTodo(todo);

  dispatch(loadTodos());
};
