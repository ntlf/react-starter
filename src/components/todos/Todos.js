import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { loadTodos } from '../../actions/todos';
import Content from '../common/Content';

export const Todos = ({ todos, loadTodos }) => {
  useEffect(() => {
    loadTodos();
  }, [loadTodos]);

  return (
    <Content>
      <div>Todos</div>

      {!todos.data.length && <div>Loading...</div>}

      {!!todos.data.length && (
        <div>
          <ul>
            {todos.data.map(todo => (
              <li key={todo.id}>{todo.text}</li>
            ))}
          </ul>
        </div>
      )}
    </Content>
  );
};

const mapStateToProps = ({ todos }) => ({
  todos,
});

const mapDispatchToProps = {
  loadTodos,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Todos);
