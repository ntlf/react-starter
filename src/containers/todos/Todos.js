/* eslint-disable react/button-has-type */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { addTodo, loadTodos } from '../../actions/todos';
import styles from './Todos.module.scss';

export class Todos extends Component {
  state = { value: '' };

  componentDidMount() {
    const { loadTodos } = this.props;

    loadTodos();
  }

  handleChange = event => {
    this.setState({ value: event.target.value });
  };

  handleSubmit = event => {
    const { addTodo } = this.props;
    const { value } = this.state;

    addTodo({
      id: Math.random()
        .toString(36)
        .substring(2, 8),
      text: value
    });

    this.setState({ value: '' });

    event.preventDefault();
  };

  render() {
    const { todos } = this.props;
    const { value } = this.state;

    if (todos.error) {
      return <div>{todos.error}</div>;
    }

    return (
      <div className={styles.Home}>
        <form onSubmit={this.handleSubmit}>
          <input type="text" value={value} onChange={this.handleChange} />
          <button>Send</button>
        </form>

        {todos.isFetching && <div>Loading...</div>}
        {todos.data &&
          !todos.isFetching && (
            <div>
              <ul>
                {todos.data.map(todo => (
                  <li key={todo.id}>{todo.text}</li>
                ))}
              </ul>
            </div>
          )}
      </div>
    );
  }
}

Todos.propTypes = {
  loadTodos: PropTypes.func.isRequired,
  addTodo: PropTypes.func.isRequired,
  todos: PropTypes.shape({
    data: PropTypes.array
  })
};

const mapStateToProps = ({ todos }) => ({
  todos
});

const mapDispatchToProps = {
  loadTodos,
  addTodo
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Todos);
