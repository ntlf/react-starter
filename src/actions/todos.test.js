import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { loadTodos } from './todos';

jest.mock('../api/todos');

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('todos actions', () => {
  describe('loadTodos', () => {
    it('is a function', () => {
      expect(loadTodos.constructor).toBe(Function);
    });

    it('should invoke actions correctly', () => {
      const store = mockStore({ todos: {} });

      return store.dispatch(loadTodos()).then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
    });
  });
});
