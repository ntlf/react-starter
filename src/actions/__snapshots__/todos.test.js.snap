// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todos actions loadTodos should invoke actions correctly 1`] = `
Array [
  Object {
    "type": "FETCH_TODOS_REQUEST",
  },
  Object {
    "payload": Object {
      "data": Array [
        Object {
          "id": "1",
          "text": "One",
        },
        Object {
          "id": "2",
          "text": "Two",
        },
      ],
    },
    "type": "FETCH_TODOS_SUCCESS",
  },
]
`;
